pkg/block/fetcher.go

```diff
@@ -0,0 +1,410 @@
+			}
+		}()
+	}
+
+	// Workers scheduled, distribute blocks.
+	err = s.bkt.Iter(ctx, "", func(name string) error {
+		id, ok := IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case ch <- id:
+		}
+
+		return nil
+	})
+	close(ch)
+
+	wg.Wait()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "MetaFetcher: iter bucket")
+	}
+
+	incompleteView := len(metaErrs) > 0
+
+	// Only for complete view of blocks update the cache.
+	if !incompleteView {
+		cached := make(map[ulid.ULID]*metadata.Meta, len(metas))
+		for id, m := range metas {
+			cached[id] = m
+		}
+		s.cached = cached
+
+		// Best effort cleanup of disk-cached metas.
+		if s.cacheDir != "" {
+			names, err := fileutil.ReadDir(s.cacheDir)
+			if err != nil {
+				level.Warn(s.logger).Log("msg", "best effort remove of not needed cached dirs failed; ignoring", "err", err)
+			} else {
+				for _, n := range names {
+					id, ok := IsBlockDir(n)
+					if !ok {
+						continue
+					}
+
+					if _, ok := metas[id]; ok {
+						continue
+					}
+
+					cachedBlockDir := filepath.Join(s.cacheDir, id.String())
+
+					// No such block loaded, remove the local dir.
+					if err := os.RemoveAll(cachedBlockDir); err != nil {
+						level.Warn(s.logger).Log("msg", "best effort remove of not needed cached dir failed; ignoring", "dir", cachedBlockDir, "err", err)
+					}
+				}
+			}
+		}
+	}
+
+	for _, f := range s.filters {
+		// NOTE: filter can update synced metric accordingly to the reason of the exclude.
+		f(metas, s.metrics.synced, incompleteView)
+	}
+
+	s.metrics.synced.WithLabelValues(loadedMeta).Set(float64(len(metas)))
+	s.metrics.synced.Submit()
+
+	if incompleteView {
+		return metas, partial, errors.Wrap(metaErrs, "incomplete view")
+	}
+
+	level.Info(s.logger).Log("msg", "successfully fetched block metadata", "duration", time.Since(start).String(), "cached", len(s.cached), "returned", len(metas), "partial", len(partial))
+	return metas, partial, nil
+}
+
+var _ MetaFetcherFilter = (&TimePartitionMetaFilter{}).Filter
+
+// TimePartitionMetaFilter is a MetaFetcher filter that filters out blocks that are outside of specified time range.
+type TimePartitionMetaFilter struct {
+	minTime, maxTime model.TimeOrDurationValue
+}
+
+// NewTimePartitionMetaFilter creates TimePartitionMetaFilter.
+func NewTimePartitionMetaFilter(MinTime, MaxTime model.TimeOrDurationValue) *TimePartitionMetaFilter {
+	return &TimePartitionMetaFilter{minTime: MinTime, maxTime: MaxTime}
+}
+
+// Filter filters out blocks that are outside of specified time range.
+func (f *TimePartitionMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) {
+	for id, m := range metas {
+		if m.MaxTime >= f.minTime.PrometheusTimestamp() && m.MinTime <= f.maxTime.PrometheusTimestamp() {
+			continue
+		}
+		synced.WithLabelValues(timeExcludedMeta).Inc()
+		delete(metas, id)
+	}
+}
+
+var _ MetaFetcherFilter = (&LabelShardedMetaFilter{}).Filter
```

Comment:
Also lets not forget to remove this and the other one. I presume some linter was complaining 😄

Reply:
Why removing this? (: This is ensuring Filter meets interface

Reply:
Ah, right, that's ok, I guess :+1: